openedx/core/djangoapps/user_authn/utils.py

```diff
@@ -76,19 +76,23 @@ def generate_username_suggestions(username):
     """ Generate 3 available username suggestions """
     max_length = USERNAME_MAX_LENGTH
     short_username = username[:max_length - 4] if max_length is not None else username
+    short_username = short_username.replace('_', '').replace('-', '')
 
     username_suggestions = []
-    int_ranges = {
-        1: {'min': 0, 'max': 9},
-        2: {'min': 10, 'max': 99},
-        3: {'min': 100, 'max': 999},
-    }
-    while len(username_suggestions) < 3:
-        int_length = len(username_suggestions) + 1
-        int_range = int_ranges[int_length]
-        random_int = random.randint(int_range['min'], int_range['max'])
-        username = f'{short_username}_{random_int}'
-        if not username_exists_or_retired(username):
-            username_suggestions.append(username)
+    int_ranges = [
+        {'min': 0, 'max': 9},
+        {'min': 10, 'max': 99},
+        {'min': 100, 'max': 999},
+        {'min': 1000, 'max': 99999},
+    ]
+    for int_range in int_ranges:
+        for int_value in range(int_range['min'], int_range['max']):
+            username_suggestion = f'{short_username}_{int_value}'
+            if not username_exists_or_retired(username_suggestion):
```

Comment:
I think this call is too expensive to make like this for the later int_ranges (the third one would result in 900 checks!), maybe we could combine the approaches with something like this:

```python
for int_range in int_ranges:
    for _ in range(min(int_range['max'], 100)):
            random_int = random.randint(int_range['min'], int_range['max'])
            username = f'{short_username}_{random_int}'
            if not username_exists_or_retired(username):
                username_suggestions.append(username)
                break
    if len(username_suggestions) == 3:
        break
```

Reply:
@mulby wouldn't it only happen if all other 899 numbers are already taken for a single username? e.g. audit_100, audit_101,...., audit_998.

I believe it will happen very rarely but I'll look into your suggestion.

Reply:
yes - I would like to avoid this scenario even if rare.

I'd rather bias toward generating fewer (or no) suggestions if we can't relatively quickly and cheaply generate them instead of spending a lot of resources on finding one.